R/sis.R

```diff
@@ -0,0 +1,150 @@
+#'   }
+#'   \item{`r_eff`}{
+#'     If specified, the user's `r_eff` argument.
+#'   }
+#'   \item{`tail_len`}{
+#'     Not used for `sis`.
+#'   }
+#'   \item{`dims`}{
+#'     Integer vector of length 2 containing `S` (posterior sample size)
+#'     and `N` (number of observations).
+#'   }
+#'   \item{`method`}{
+#'     Method used for importance sampling, here `sis`.
+#'   }
+#' }
+#'
+#' @seealso
+#' \itemize{
+#' \item [psis()] for approximate LOO-CV using PSIS.
+#' \item [loo()] for approximate LOO-CV.
+#' \item [pareto-k-diagnostic] for PSIS diagnostics.
+#' }
+#'
+#' @template loo-and-psis-references
+#'
+#' @examples
+#' log_ratios <- -1 * example_loglik_array()
+#' r_eff <- relative_eff(exp(-log_ratios))
+#' sis_result <- sis(log_ratios, r_eff = r_eff)
+#' str(sis_result)
+#'
+#' # extract smoothed weights
+#' lw <- weights(sis_result) # default args are log=TRUE, normalize=TRUE
+#' ulw <- weights(sis_result, normalize=FALSE) # unnormalized log-weights
+#'
+#' w <- weights(sis_result, log=FALSE) # normalized weights (not log-weights)
+#' uw <- weights(sis_result, log=FALSE, normalize = FALSE) # unnormalized weights
+#'
+#' @export
+sis <- function(log_ratios, ...) UseMethod("sis")
+
+#' @export
+#' @templateVar fn sis
+#' @template array
+#'
+sis.array <-
+  function(log_ratios, ...,
+           r_eff = NULL,
+           cores = getOption("mc.cores", 1)) {
+  importance_sampling.array(log_ratios = log_ratios, ...,
+                            r_eff = r_eff,
+                            cores = cores,
+                            method = "sis")
+  }
+
+#' @export
+#' @templateVar fn sis
+#' @template matrix
+#'
+sis.matrix <-
+  function(log_ratios,
+           ...,
+           r_eff = NULL,
+           cores = getOption("mc.cores", 1)) {
+    importance_sampling.matrix(log_ratios,
+                               ...,
+                               r_eff = r_eff,
+                               cores = cores,
+                               method = "sis")
+  }
+
+#' @export
+#' @templateVar fn sis
+#' @template vector
+#'
+sis.default <-
+  function(log_ratios, ..., r_eff = NULL) {
+    importance_sampling.default(log_ratios = log_ratios, ...,
+                                r_eff = r_eff,
+                                method = "sis")
+  }
+
+#' @rdname psis
+#' @export
+is.sis <- function(x) {
+  inherits(x, "sis") && is.list(x)
+}
+
+
+# internal ----------------------------------------------------------------
+
+
+#' Standard IS on a single vector
+#'
+#' @noRd
+#' @param log_ratios_i A vector of log importance ratios (for `loo()`, negative
+#'   log likelihoods).
+#' @param tail_len_i Not used. Included to conform to PSIS API.
+#'
+#' @details Implementation standard importance sampling.
+#' @return A named list containing:
+#' * `lw`: vector of unnormalized log weights
+#' * `pareto_k`: scalar Pareto k estimate. For IS, this defaults to 0.
+do_sis_i <- function(log_ratios_i, tail_len_i) {
```

Comment:
Where is `tail_len_i` used in this function? I don't see it here.

Reply:
It is not used. It comes from the fact that `tail_len_i` is supplied by default to any `do_XXis_i` function. Maybe state that this as a comment in the code?

Reply:
Can we replace it by `...` so that anything may be passed but is then simply ignored? This may be a cleaner approach than having an argument which actually takes no effect.

Reply:
I fully agree. I'll see if that solves the problem.

Reply:
Seem to work.